database/migrations/2021_08_06_184400_refactor_album_model.php

```diff
@@ -0,0 +1,823 @@
+			$this->dropForeignIfExist($table, 'size_variants_photo_id_foreign');
+			$this->dropUniqueIfExist($table, 'size_variants_photo_id_size_variant_unique');
+		});
+		Schema::rename('size_variants', 'size_variants_tmp');
+		Schema::drop('sym_links');
+	}
+
+	/**
+	 * Drops temporary tables which have been created by
+	 * {@link RefactorAlbumModel::renameTables()}.
+	 *
+	 * The order is important to avoid error due to unsatisfied foreign
+	 * constraints.
+	 */
+	private function dropTemporarilyRenamedTables(): void
+	{
+		// We must remove any foreign link from `albums` to `photos` to
+		// break up circular dependencies.
+		DB::table('albums_tmp')->update(['cover_id' => null]);
+		Schema::drop('size_variants_tmp');
+		Schema::drop('photos_tmp');
+		Schema::drop('albums_tmp');
+	}
+
+	/**
+	 * Creates the table `base_albums`.
+	 *
+	 * The table `base_albums` contains all columns of the old table
+	 * `albums` which are common to normal albums and tag albums.
+	 */
+	private function createBaseAlbumTable(): void
+	{
+		Schema::create('base_albums', function (Blueprint $table) {
+			// Column definitions
+			$table->bigIncrements('id')->nullable(false);
+			$table->dateTime('created_at')->nullable(false);
+			$table->dateTime('updated_at')->nullable(false);
+			$table->string('title', 100)->nullable(false);
+			$table->text('description')->nullable();
+			$table->unsignedInteger('owner_id')->nullable(false)->default(0);
+			$table->boolean('is_public')->nullable(false)->default(false);
+			$table->boolean('grants_full_photo')->nullable(false)->default(true);
+			$table->boolean('requires_link')->nullable(false)->default(false);
+			$table->boolean('is_downloadable')->nullable(false)->default(false);
+			$table->boolean('is_share_button_visible')->nullable(false)->default(false);
+			$table->boolean('is_nsfw')->nullable(false)->default(false);
+			$table->string('password', 100)->nullable()->default(null);
+			$table->string('sorting_col', 30)->nullable()->default(null);
+			$table->string('sorting_order', 4)->nullable()->default(null);
+			// Indices and constraint definitions
+			$table->foreign('owner_id')->references('id')->on('users');
+			$table->index('is_public');
+		});
+	}
+
+	/**
+	 * Creates the table `albums` acc. to the new architecture.
+	 *
+	 * The new table `albums` only contains the columns which are specific
+	 * to real albums and are irrelevant for tag albums.
+	 */
+	private function createAlbumTableUp(): void
+	{
+		Schema::create('albums', function (Blueprint $table) {
+			// Column definitions
+			$table->unsignedBigInteger('id')->nullable(false);
+			$table->unsignedBigInteger('parent_id')->nullable()->default(null);
+			$table->string('license', 20)->nullable(false)->default('none');
+			$table->unsignedBigInteger('cover_id')->nullable()->default(null);
+			$table->unsignedBigInteger('_lft')->nullable(false)->default(0);
+			$table->unsignedBigInteger('_rgt')->nullable(false)->default(0);
```

Comment:
So many questions at once:

> Are they [`_lft` and `_rgt`] only important for albums that support sub-albums?

No, they are also important for albums which are sub-albums themselves.

> Mind you, it's not my intention to request that we immediately and fully support it within this PR, but rather that we future-proof the DB, especially given the problems with SQLite that you documented here...

This would make things extremely complicated now. It does not suffice to simply move `parent_id`, `_lft` and `_rgt` to `base_albums` but I also need to check with `ForwardsToParentImpl` if everything works fine. If we really want to make tag albums more useful this must be done thoroughly.

> So based on the discussion we've had in #1123

As you know I would like to get rid of tag albums entirely. I wanted to already have written down some of my reasons by now, but haven't found the time. I also noticed your counter arguments in #1123 and I actually was hoping to hear those arguments, because I had have a good counter-argument for each of them up my sleeve. However, I did not wanted to further drive the discussion in #1123, because this is not were the discussion should take place. I going to write on Gitter soon. We should really discuss this.

> On a less important note, I'm wondering about `cover_id` for tag albums as well.

I totally understand why someone wants to have this for tag albums. But we must also consider what is supposed to haven, if one picks an photo as a cover for a tag album and at some later point of time that photo is not anymore part of the tag album, because the tags were edited? This is not as easy as it might seem and needs thorough discussion as well. (One more argument, why I don't like tag albums.)

Reply:
> I did not wanted to further drive the discussion in #1123, because this is not were the discussion should take place. I going to write on Gitter soon. We should really discuss this.

Yes, agreed. A random issue is not a good place for such a conversation.

> But we must also consider what is supposed to haven, if one picks an photo as a cover for a tag album and at some later point of time that photo is not anymore part of the tag album, because the tags were edited? This is not as easy as it might seem and needs thorough discussion as well. (One more argument, why I don't like tag albums.)

All perfectly rational arguments, but here's a twist: we _already_ face most of these issues for regular albums as well. The server implementation allows the user to set any visible photo as a cover; I don't believe that we ever check if it's even a photo from that album, and we certainly don't reset it when the cover photo is moved/deleted/etc. Heck, once set, we don't even check if the photo is visible to the viewer; if it exists, we display it (i.e., a photo from a private subalbum). We simply agreed that doing it correctly would be "too much work", and that what we had was good enough since it's the user who explicitly sets the cover, so if they choose something they shouldn't, well, they have only themselves to blame :wink:.

Reply:
Yes, I noticed that for real albums as well. However, with the new architecture and some minor add-ons, this can easily be done right now.